src/DotNetty.Handlers/Tls/TlsHandler.cs

```diff
@@ -280,61 +521,93 @@ public override void Flush(IChannelHandlerContext context)
 
             if (!this.EnsureAuthenticated())
             {
-                this.state |= State.FlushedBeforeHandshake;
+                this.state |= TlsHandlerState.FlushedBeforeHandshake;
                 return;
             }
 
-            this.Wrap(context);
-            context.Flush();
+            try
+            {
+                this.Wrap(context);
+            }
+            finally
+            {
+                // We may have written some parts of data before an exception was thrown so ensure we always flush.
+                context.Flush();
+            }
         }
 
         void Wrap(IChannelHandlerContext context)
         {
             Contract.Assert(context == this.capturedContext);
 
-            while (true)
+            IByteBuffer buf = null;
+            try
             {
-                object msg = this.pendingUnencryptedWrites.Current;
-                if (msg == null)
+                while (true)
                 {
-                    break;
-                }
+                    List<object> messages = this.pendingUnencryptedWrites.Current;
+                    if (messages == null || messages.Count == 0)
+                    {
+                        break;
+                    }
 
-                var buf = (IByteBuffer)msg;
-                buf.ReadBytes(this.sslStream, buf.ReadableBytes); // this leads to FinishWrap being called 0+ times
+                    if (messages.Count == 1)
+                    {
+                        buf = (IByteBuffer)messages[0];
+                    }
+                    else
+                    {
+                        buf = context.Allocator.Buffer((int)this.pendingUnencryptedWrites.CurrentSize);
```

Comment:
not to package every single buffer as a separate SSL record. Every call to `SslStream.Write(...)` results in a new SSL record.